text/0000-tracked-properties.md

```diff
@@ -227,34 +248,350 @@ memoization) if a property shows up as a bottleneck during profiling.
 
 ### Native Classes
 
-Tracked properties are designed from the ground up for native JavaScript (ES6) classes.
-
-### "Just JavaScript"
-
+Tracked properties are designed from the ground up for native JavaScript (ES6)
+classes. Adopting them will make it easier to convert away from the aging Ember
+object-model, and
 
 ## Detailed design
 
 ### Autotracking Stack
-> This is the bulk of the RFC.
 
-> Explain the design in enough detail for somebody
-familiar with the framework to understand, and for somebody familiar with the
-implementation to implement. This should get into specifics and corner-cases,
-and include examples of how the feature is used. Any new terminology should be
-defined here.
+Autotracking works by taking advantage of the fact that all mutable properties
+will be decorated by `tracked`. This decorator adds a getter and a setter that
+wraps the property, be it a tracked getter or a tracked simple property. Every
+time we activate a tracked getter, it creates a global stack before it begins
+executing the getter's code, and as it executes, any other tracked wrappers
+that run add themselves to the stack.
+
+This example demonstrates how this method works, but is simplified to only work
+with tracked getters, and to only track keys. It uses the stage 2 decorator
+syntax.
+
+```js
+let PROPERTY_STACK;
+
+function tracked({ key, descriptor }) {
+  let originalGet = descriptor.get;
+
+  descriptor.get = function() {
+    // store the stack from the previous getter, if it exists
+    let previousStack = PROPERTY_STACK;
+
+    // create a new stack
+    PROPERTY_STACK = [];
+
+    // call the original get function to get the value
+    let value = originalGet.call(this);
+
+    // At this point, PROPERTY_STACK will contain all of the keys of any values
+    // that were marked as tracked and accessed when `originalGet` was called.
+    // We can use this knowledge to setup our change tracking, then push our
+    // own key and the current stack into the previous stack if it exists.
+    if (previousStack) {
+      previousStack.push(key, ...PROPERTY_STACK);
+    }
+
+    PROPERTY_STACK = previousStack;
+
+    return value;
+  }
+}
+```
+
+Of course, tracking keys would only work if we knew the full paths, or were only
+concerned with only changes to the current object. The real implementation uses
+[Glimmer VM's internal validators](https://github.com/glimmerjs/glimmer-vm/blob/master/guides/05-validators.md)
+to track changes to properties.
+
+As properties are accessed, their validation  _tags_ are pushed on the stack,
```

Comment:
I'm worried that this section goes into too much detail, and opens up topics that would be difficult to cover properly in this RFC. Specifically, I'm worried that too much implementation detail will cause people to be more confused, vs. higher level language that captures the overall idea without requiring the reader to understand tags/validators etc.

That said, it's understandable that people will be skeptical here if it seems like too much magic. Would love your thoughts on how to strike the right balance.

Reply:
Or maybe we just do need to go into a deeper level of detail in other sections, specifically covering tags, invalidations, etc.